rust/pattern/constraint/type_constraint.rs

```diff
@@ -0,0 +1,273 @@
+    }
+}
+
+impl From<(&str, &str)> for PlaysConstraint {
+    fn from((relation_type, role_type): (&str, &str)) -> Self {
+        PlaysConstraint::from((String::from(relation_type), String::from(role_type)))
+    }
+}
+
+impl From<(String, String)> for PlaysConstraint {
+    fn from((relation_type, role_type): (String, String)) -> Self {
+        PlaysConstraint::from(ScopedType::from((relation_type, role_type)))
+    }
+}
+
+impl From<ScopedType> for PlaysConstraint {
+    fn from(scoped_type: ScopedType) -> Self {
+        PlaysConstraint::new(
+            UnboundVariable::hidden().type_(scoped_type).unwrap().into_type(),
+            None,
+        )
+    }
+}
+
+impl From<UnboundVariable> for PlaysConstraint {
+    fn from(role_type: UnboundVariable) -> Self {
+        PlaysConstraint::from(role_type.into_type())
+    }
+}
+impl From<TypeVariable> for PlaysConstraint {
+    fn from(role_type: TypeVariable) -> Self {
+        PlaysConstraint::new(role_type, None)
+    }
+}
+
+impl Display for PlaysConstraint {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "plays {}", self.role_type)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum IsKey {
+    Yes,
+    No,
+}
+
+impl Display for IsKey {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if *self == IsKey::Yes {
+            f.write_str(" @key")?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct OwnsConstraint {
+    pub attribute_type: TypeVariable,
+    pub overridden_attribute_type: Option<TypeVariable>,
+    pub is_key: IsKey,
+}
+
+impl OwnsConstraint {
+    fn new(
+        attribute_type: TypeVariable,
+        overridden_attribute_type: Option<TypeVariable>,
+        is_key: IsKey,
+    ) -> Self {
+        OwnsConstraint { attribute_type, overridden_attribute_type, is_key }
+    }
+}
+
+impl From<&str> for OwnsConstraint {
+    fn from(attribute_type: &str) -> Self {
+        OwnsConstraint::from((attribute_type, IsKey::No))
+    }
+}
+
+impl From<(&str, IsKey)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (&str, IsKey)) -> Self {
+        OwnsConstraint::from((
+            UnboundVariable::hidden().type_(attribute_type).unwrap().into_type(),
+            is_key,
+        ))
+    }
+}
+impl From<(String, IsKey)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (String, IsKey)) -> Self {
+        OwnsConstraint::from((
+            UnboundVariable::hidden().type_(attribute_type).unwrap().into_type(),
+            is_key,
+        ))
+    }
+}
+impl From<(UnboundVariable, IsKey)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (UnboundVariable, IsKey)) -> Self {
+        OwnsConstraint::from((attribute_type.into_type(), is_key))
+    }
+}
+impl From<(TypeVariable, IsKey)> for OwnsConstraint {
+    fn from((attribute_type, is_key): (TypeVariable, IsKey)) -> Self {
+        OwnsConstraint::new(attribute_type, None, is_key)
+    }
+}
```

Comment:
`owns((var("a"), IsKey::Yes))`

Yes, some input combinations seem to be missing.